### PR TITLE
[WIP] feat(wandb): add logging to TFTrainer

### PR DESCRIPTION
**Pull request by borisdayma**
Bring logging feature parity from `Trainer` to `TFTrainer`.
Code has been refactored to share logging utilities.

### CODE REVIEW

**Comment by borisdayma** [comment]
Note that there are still a few differences.

For example `TFTrainer` uses `args.eval_steps`.
It could make sense to share refactor training args so they share the same ones when possible.

**Comment by borisdayma** [comment]
Here is an example run with `TFTrainer` on MRPC.

image

[Link to W&B run](https://app.wandb.ai/borisd13/huggingface/runs/1zngxzw0?workspace=user-borisd13)

File: src/transformers/trainer_utils.py [hunk 1]
```diff
@@ -1,6 +1,83 @@
...
+                for k, v in logs.items():
+                    tf.summary.scalar(k, v, step=trainer.global_step)
+        trainer.tb_writer.flush()
+    if is_wandb_available():
+        trainer._wandb.log(logs, step=trainer.global_step)
+    output = {**logs, **{"step": trainer.global_step}}
+    if iterator is not None:
+        iterator.write(output)
+    else:
+        print(output)
```
Comment (jplu) [inline src/transformers/trainer_utils.py:80]: The logger should be used here instead of a print.

File: src/transformers/trainer_utils.py [hunk 2]
```diff
@@ -1,6 +1,83 @@
...
+    if trainer.tb_writer:
+        if trainer.__class__.__name__ == "Trainer":
+            for k, v in logs.items():
+                trainer.tb_writer.add_scalar(k, v, trainer.global_step)
+        elif trainer.__class__.__name__ == "TFTrainer":
+            import tensorflow as tf
+
+            with trainer.tb_writer.as_default():
+                for k, v in logs.items():
+                    tf.summary.scalar(k, v, step=trainer.global_step)
```
Comment (jplu) [inline src/transformers/trainer_utils.py:72]: IMO these parts should belong to their respective Trainer.

File: src/transformers/trainer_utils.py [hunk 3]
```diff
@@ -1,6 +1,83 @@
...
+    # keep track of model topology and gradients (only with Pytorch)
+    if os.getenv("WANDB_WATCH") != "false" and trainer.__class__.__name__ == "Trainer":
+        wandb.watch(
+            trainer.model, log=os.getenv("WANDB_WATCH", "gradients"), log_freq=max(100, trainer.args.logging_steps)
+        )
+    # give access to wandb module
+    trainer._wandb = wandb
+
+
+def log_metrics(trainer, logs: Dict[str, float], iterator: Optional[tqdm] = None) -> None:
```
Comment (jplu) [inline src/transformers/trainer_utils.py:60]: Epoch and step might be passed as arguments which doesn't make too changes in the Trainer classes.

File: src/transformers/trainer_tf.py [hunk 4]
```diff
@@ -273,45 +284,44 @@ def train(self) -> None:
         logger.info("  Num Epochs = %d", epochs)
         logger.info("  Total optimization steps = %d", self.train_steps)
 
+        self.global_step = 0
+        self.epoch = 0
```
Comment (jplu) [inline src/transformers/trainer_tf.py:288]: Can you elaborate more on why this variable plz? Why not taking the epoch value just above.

File: src/transformers/trainer_tf.py [hunk 5]
```diff
@@ -273,45 +284,44 @@ def train(self) -> None:
         logger.info("  Num Epochs = %d", epochs)
         logger.info("  Total optimization steps = %d", self.train_steps)
 
+        self.global_step = 0
```
Comment (jplu) [inline src/transformers/trainer_tf.py:287]: This variable seems a repetition with the `iterations` variable can be removed.

**Review by jplu — COMMENTED** [review]
Hello,

Thanks a lot for this PR, will be very useful!! :)

I did some comments :)

File: src/transformers/trainer_utils.py (hunk 1 above)
Comment (borisdayma) [reply src/transformers/trainer_utils.py:80]: Noted, this section comes from `Trainer` but I agree that it makes more sense to use the logger.

File: src/transformers/trainer_utils.py (hunk 2 above)
Comment (borisdayma) [reply src/transformers/trainer_utils.py:72]: I see your point.
This is only related to the tensorboard logging so I could have a `trainer.tb_log` function associated to each `Trainer` and `TFTrainer` class.

File: src/transformers/trainer_utils.py (hunk 3 above)
Comment (borisdayma) [reply src/transformers/trainer_utils.py:60]: Here it was mainly to keep it the same way as `Trainer`. There are a few advantages in doing it this way:

* some example scripts call for evaluation after full training (ex: `run_glue`), where the epoch & step are not accessible any longer, which were the reasons why we added them within the class.
* access to `trainer.tb_log` discussed in your previous comment will also be easier this way

File: src/transformers/trainer_tf.py (hunk 4 above)
Comment (borisdayma) [reply src/transformers/trainer_tf.py:288]: This is a float value showing number of epochs elapsed.
When comparing training runs of different batch size, or different data size, you will have different number of steps. You can plot loss vs epoch and superpose multiple runs this way.

image

File: src/transformers/trainer_tf.py (hunk 5 above)
Comment (borisdayma) [reply src/transformers/trainer_tf.py:287]: This is to make it compatible with the functions moved from `Trainer` to `TFTrainer`.
Could `iterations` be removed or does it capture something more complex (such as reloading from checkpoints…).

**Comment by borisdayma** [comment]
@jplu while I'm doing it, is the `step` variable from `_prediction_loop` used or can I remove it?

File: src/transformers/trainer_utils.py (hunk 2 above)
Comment (jplu) [reply src/transformers/trainer_utils.py:72]: Very good idea!

File: src/transformers/trainer_utils.py (hunk 3 above)
Comment (jplu) [reply src/transformers/trainer_utils.py:60]: For evaluation I don't think you need the epoch. For the step yes indeed. But try to do with the `trainer.tb_log` it should be easier for me to see what you mean :)

File: src/transformers/trainer_tf.py (hunk 5 above)
Comment (jplu) [reply src/transformers/trainer_tf.py:287]: It captures the step directly from the checkpoint if any so, it is a real duplicate of global.

File: src/transformers/trainer_tf.py (hunk 4 above)
Comment (jplu) [reply src/transformers/trainer_tf.py:288]: OH I see!!! I didn't know that (sorry I'm not really familiar with wandb) it ok makes sense.

**Comment by jplu** [comment]
It is used for logging.

**Comment by jplu** [comment]
Ping @julien-c as there are some changes in the PT trainer.

**Comment by borisdayma** [comment]
> It is used for logging.

Ok, I don't see it being used anywhere in that function…

**Comment by jplu** [comment]
Hummm indeed... Can you still keep it, I will review this later to better check that part :)

**Comment by borisdayma** [comment]
I should have addressed your comments. I moved tensorboard specific logging directly within specific trainers and call it from shared `log_metrics`.
It could have also been the opposite with a public `trainer.log` method that calls `log_metrics` for everything non-tensorboard (wandb and stdout) but I thought this separation would be less obvious to users.

Let me know if you have any other suggestions.

The main differences now between the 2 trainers are the use of `args.debug` and `args.eval_steps` in `TFTrainer`.

File: src/transformers/trainer_tf.py (hunk 4 above)
Comment (jplu) [reply src/transformers/trainer_tf.py:288]: Would be nice if you can give a more elaborate name instead of just `epoch` (to do not confuse with the loop epoch variable) and add a comment explaining that it is used only for wandb.

**Review by jplu — COMMENTED** [review]
Looks much better well done :) Just again one tiny thing.

File: src/transformers/trainer_tf.py (hunk 4 above)
Comment (borisdayma) [reply src/transformers/trainer_tf.py:288]: Ok, I just kept the same name as the one used in `Trainer`.
Any suggestion? `epoch_elapsed` would also bring confusion.
Another possibility would be to change the loop epoch to `epoch_iter`. What do you think?

Comment (jplu) [reply src/transformers/trainer_tf.py:288]: I suggest `epoch_logging` and indeed `epoch_iter` changing both + a small comment for `epoch_logging`

**Comment by borisdayma** [comment]
Let me know if anything else is needed

File: src/transformers/trainer.py [hunk 6]
```diff
@@ -287,7 +287,7 @@ def get_optimizers(
         If you want to use something else, you can pass a tuple in the Trainer's init,
         or override this method in a subclass.
         """
-        if self.optimizers is not None:
+        if self.optimizers is not Noneself.eval:
```
Comment (jplu) [inline src/transformers/trainer.py]: Be careful here!

Comment (borisdayma) [reply src/transformers/trainer.py]: must have been a middle click by mistake… Shows how tests are important.
I'll probably add some loggers tests in the future in a separate PR.

**Comment by borisdayma** [comment]
The last CI error seems unrelated to this PR.

**Comment by jplu** [comment]
Ok it is fine for me for the TF part. I let @julien-c to review.

**Comment by borisdayma** [comment]
Can I have any feedback on any possible changes still required?
This is a pretty large refactor of logging so it's hard to keep up to date with the repo.

File: src/transformers/trainer_utils.py [hunk 7]
```diff
@@ -1,6 +1,79 @@
...
+    Environment:
+        WANDB_WATCH:
+            (Optional, ["gradients", "all", "false"]) "gradients" by default, set to "false" to disable gradient logging
+            or "all" to log gradients and parameters
+        WANDB_PROJECT:
+            (Optional): str - "huggingface" by default, set this to a custom string to store results in a different project
+        WANDB_DISABLED:
+            (Optional): boolean - defaults to false, set to "true" to disable wandb entirely
+    """
+    if trainer.is_world_master():
```
Comment (borisdayma) [inline src/transformers/trainer_utils.py:48]: Here we need to do one of the following:

* Option 1: add a `TFTrainer.is_world_master` method (I would need help for that)
* Option 2: replace with something like `if trainer.__class__.__name__ == "TFTrainer" or trainer.is_world_master()`

What do you think?

**Comment by borisdayma** [comment]
I merged master. Main change is logging for wandb & Tensorboard is applied only for world master now.
Need to figure out if the same should apply for `TFTrainer`. I left a comment on the code.

File: src/transformers/trainer_utils.py (hunk 7 above)
Comment (jplu) [reply src/transformers/trainer_utils.py:48]: No need of a `.is_world_master()` method in TF.

**Comment by jplu** [comment]
Thanks! Looks ok, about the world master there is no need for the TF part.

File: src/transformers/trainer_utils.py (hunk 7 above)
Comment (borisdayma) [reply src/transformers/trainer_utils.py]: Ok I moved it directly to `Trainer` so it does not apply to `TFTrainer`

File: src/transformers/trainer.py [hunk 8]
```diff
@@ -552,32 +514,24 @@ def train(self, model_path: Optional[str] = None):
...
-        else:
-            print(output)
+    def _log_tb(self, logs: Dict[str, float]) -> None:
+        """
+        Log data to tensorboard. Called by `log_metrics`.
+        """
+        for k, v in logs.items():
+            if isinstance(v, (int, float)):
+                self.tb_writer.add_scalar(k, v, self.global_step)
+            else:
```
Comment (jplu) [inline src/transformers/trainer.py:524]: In case the value is a string.

```suggestion
            elif isinstance(v, str):
                tf.summary.text(k, v, self.global_step)
            else:
```
Or even better you can cast the value to `str` to log it and log the error in case it cannot be cast.

File: src/transformers/trainer.py [hunk 9]
```diff
@@ -713,7 +667,8 @@ def evaluate(
 
         output = self._prediction_loop(eval_dataloader, description="Evaluation")
 
-        self._log(output.metrics)
+        if self.is_world_master:
```
Comment (jplu) [inline src/transformers/trainer.py:670]: To be removed.

File: src/transformers/trainer_utils.py [hunk 10]
```diff
@@ -1,6 +1,77 @@
...
+        WANDB_PROJECT:
+            (Optional): str - "huggingface" by default, set this to a custom string to store results in a different project
+        WANDB_DISABLED:
+            (Optional): boolean - defaults to false, set to "true" to disable wandb entirely
+    """
+    logger.info('Automatic Weights & Biases logging enabled, to disable set os.environ["WANDB_DISABLED"] = "true"')
+
+    wandb.init(project=os.getenv("WANDB_PROJECT", "huggingface"), config=vars(trainer.args))
+    # keep track of model topology and gradients (only with Pytorch)
+    if os.getenv("WANDB_WATCH") != "false" and trainer.__class__.__name__ == "Trainer":
```
Comment (jplu) [inline src/transformers/trainer_utils.py:52]: Here instead of checking if the call comes from the PT or TF you should put this part into the PT trainer as it is not available in TF.

File: src/transformers/trainer.py [hunk 11]
```diff
@@ -426,8 +387,8 @@ def train(self, model_path: Optional[str] = None):
         logger.info("  Gradient Accumulation steps = %d", self.args.gradient_accumulation_steps)
         logger.info("  Total optimization steps = %d", t_total)
 
-        self.global_step = 0
-        self.epoch = 0
+        self.global_step = 0  # used by loggers
+        self.epoch_logging = 0  # added with loggers to use as x-axis
```
Comment (julien-c) [inline src/transformers/trainer.py:391]: nit: can you add comments on their own line (only if they're really needed)

File: src/transformers/trainer_utils.py [hunk 12]
```diff
@@ -1,6 +1,77 @@
...
+        wandb.watch(
+            trainer.model, log=os.getenv("WANDB_WATCH", "gradients"), log_freq=max(100, trainer.args.logging_steps)
+        )
+    # give access to wandb module
+    trainer._wandb = wandb
```
Comment (jplu) [inline src/transformers/trainer_utils.py:57]: Better to return `wandb` than to set it.

File: src/transformers/trainer_tf.py [hunk 13]
```diff
@@ -265,44 +285,40 @@ def train(self) -> None:
         logger.info("  Num Epochs = %d", epochs)
         logger.info("  Total optimization steps = %d", self.train_steps)
 
-        for epoch in range(start_epoch, int(epochs + 1)):
-            for training_loss in self._training_steps(train_ds, optimizer):
-                step = iterations.numpy()
+        self.global_step = 0  # used by loggers
+        self.epoch_logging = 0  # added with loggers to use as x-axis
```
Comment (julien-c) [inline src/transformers/trainer_tf.py:289]: same here

File: src/transformers/trainer_tf.py [hunk 14]
```diff
@@ -265,44 +285,40 @@ def train(self) -> None:
         logger.info("  Num Epochs = %d", epochs)
         logger.info("  Total optimization steps = %d", self.train_steps)
 
-        for epoch in range(start_epoch, int(epochs + 1)):
-            for training_loss in self._training_steps(train_ds, optimizer):
-                step = iterations.numpy()
+        self.global_step = 0  # used by loggers
```
Comment (jplu) [inline src/transformers/trainer_tf.py:288]: The global init here can even be removed.

File: src/transformers/trainer_utils.py [hunk 15]
```diff
@@ -1,6 +1,77 @@
...
+
+
+def log_metrics(trainer, logs: Dict[str, float], iterator: Optional[tqdm] = None) -> None:
```
Comment (julien-c) [inline src/transformers/trainer_utils.py:60]: Not sure we want to share that accross both Trainer to be honest, this seems weird

File: src/transformers/trainer_utils.py [hunk 16]
```diff
@@ -1,6 +1,77 @@
...
+    """
+    Log metrics with available loggers.
+    """
+    if trainer.epoch_logging is not None:
+        logs["epoch"] = trainer.epoch_logging
+    if trainer.tb_writer:
+        trainer._log_tb(logs)
+    if is_wandb_available():
+        trainer._wandb.log(logs, step=trainer.global_step)
```
Comment (julien-c) [inline src/transformers/trainer_utils.py:69]: can't we just do `wandb.log`?

File: src/transformers/trainer_utils.py (hunk 15 above)
Comment (borisdayma) [reply src/transformers/trainer_utils.py:60]: Are you talking about both the setup and logging part?

Initially I didn't notice tensorboard logging would be different between the 2 methods but now it may make sense to keep entire logging separate.

What about setting up wandb (I would just add a bool "watch" argument to make it clearer).

File: src/transformers/trainer_tf.py (hunk 14 above)
Comment (borisdayma) [reply src/transformers/trainer_tf.py:288]: It is used if you want to eval several pre-trained models without additional training.

I can also change the logging logic to check if this parameter is present at all.

Comment (jplu) [reply src/transformers/trainer_tf.py:288]: We already set it to 0 in the ``__init__`

**Comment by borisdayma** [comment]
Based on above comments, I can just put all the wandb logic separately in their respective Trainer's (and not use `trainer_utils` anymore).

Could you confirm this is the way to go @jplu @julien-c

**Comment by borisdayma** [comment]
I made a new PR as I understood you prefer not to refactor logging into a single file like I did here.
Feel free to close this one if my understanding was correct.

**Comment by borisdayma** [comment]
Obsolete PR
